test/integration/otel_test.go

```diff
@@ -0,0 +1,295 @@
+
+package integration
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/eggsampler/acme/v3"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/letsencrypt/boulder/cmd"
+	blog "github.com/letsencrypt/boulder/log"
+	"github.com/letsencrypt/boulder/test"
+)
+
+type SpanData struct {
+	SpanID        string
+	OperationName string
+	Warnings      []string
+	ProcessID     string
+	References    []struct {
+		RefType string
+		TraceID string
+		SpanID  string
+	}
+}
+
+type TraceData struct {
+	TraceID   string
+	Spans     []SpanData
+	Processes map[string]struct {
+		ServiceName string
+	}
+	Warnings []string
+}
+
+// TraceResponse is what we get from the traces API.
+// We always search for a single trace by ID, so this should be length 1.
+type TraceResponse struct {
+	Data []TraceData
+}
+
+func getTraceFromJaeger(t *testing.T, traceID trace.TraceID) TraceData {
+	traceURL := "http://bjaeger:16686/api/traces/" + traceID.String()
+	resp, err := http.Get(traceURL)
+	test.AssertNotError(t, err, "failed to trace from jaeger: "+traceID.String())
+	if resp.StatusCode == http.StatusNotFound {
+		t.Fatalf("jaeger returned 404 for trace %s", traceID)
+	}
+	test.AssertEquals(t, resp.StatusCode, http.StatusOK)
+
+	body, err := io.ReadAll(resp.Body)
+	test.AssertNotError(t, err, "failed to read trace body")
+
+	var parsed TraceResponse
+	err = json.Unmarshal(body, &parsed)
+	test.AssertNotError(t, err, "failed to decode traces body")
+
+	if len(parsed.Data) != 1 {
+		t.Fatalf("expected to get exactly one trace for %s: %v", traceID, parsed)
+	}
+
+	return parsed.Data[0]
+}
+
+type expectedSpans struct {
+	Operation string
+	Service   string
+	Children  []expectedSpans
+}
+
+// isParent returns true if the given span has a parent of ParentID
+// The empty string means no ParentID
+func isParent(parentID string, span SpanData) bool {
+	if len(span.References) == 0 {
+		return parentID == ""
+	}
+	for _, ref := range span.References {
+		// In OpenTelemetry, CHILD_OF is the only reference, but Jaeger supports other systems.
+		if ref.RefType == "CHILD_OF" {
+			return ref.SpanID == parentID
+		}
+	}
+	return false
+}
+
+func missingChildren(traceData TraceData, spanID string, children []expectedSpans) bool {
+	for _, child := range children {
+		if !findSpans(traceData, spanID, child) {
+			// Missing Child
+			return true
+		}
+	}
+	return false
+}
+
+// findSpans checks if the expectedSpan and its expected children are found in traceData
+func findSpans(traceData TraceData, parentSpan string, expectedSpan expectedSpans) bool {
+	for _, span := range traceData.Spans {
+		if !isParent(parentSpan, span) {
+			continue
+		}
+		if expectedSpan.Service != traceData.Processes[span.ProcessID].ServiceName {
```

Comment:
This is a super-nit, but I find it's generally more readable if the "real" value goes on the left, and the "expected" value goes on the right -- e.g. we don't say `if 2 == myVar`.

Same comment on line 120.